src/utils/utils.jl

```diff
@@ -87,25 +87,78 @@ function strip_parametric_type(name::AbstractString)
 end
 
 """
-Return a Tuple of type and parameter types for cases where a parametric type has been
-encoded as a string. If the type is not parameterized then just return the type.
+    parse_serialized_type(type_str::AbstractString, mod::Module)
+
+Return a symbol parsed from a serialized string. Sanitizes the input first and attempts to
+ensure that the string does not contain code. The caller can call `eval` on the returned
+symbol.
+
+Throws ErrorException if the string contains invalid characters or code.
+
 """
-function separate_type_and_parameter_types(name::String)
-    parameters = Vector{String}()
-    index_start_brace = findfirst("{", name)
-    if isnothing(index_start_brace)
-        type_str = name
+function parse_serialized_type(serialized_type::Union{Symbol, String})
+    if serialized_type isa Symbol
+        serialized_type = string(serialized_type)
+    end
+
+    _check_expression_characters(serialized_type)
+    val = Meta.parse(serialized_type)
+    if val isa Symbol
+        return val
+    elseif val isa Expr
+        _check_parametric_expression(val)
     else
-        type_str = name[1:(index_start_brace.start - 1)]
-        index_close_brace = findfirst("}", name)
-        @assert index_start_brace.start < index_close_brace.start
-        for x in
-            split(name[(index_start_brace.start + 1):(index_close_brace.start - 1)], ",")
-            push!(parameters, strip(x))
+        error("invalid type: $val")
+    end
+
+    return val
+end
+
+function _check_expression_characters(serialized_type::String)
+    # This will return false if the char is any symbol or operator other than
+    # '{', '}', and ',', which should be the only non-alphanumeric characters allowed
+    # in a type that we are handling during serialization (as of now).
+    # Look up an ASCII table to interpret these.
+    for char in serialized_type
+        is_valid = false
+        value = codepoint(char)
+        if value < 32
+            is_valid = false
+        elseif value >= 33 && value <= 47 && value != 44
+            is_valid = false
+        elseif value >= 58 && value <= 64
+            is_valid = false
+        elseif value >= 91 && value <= 94
+            is_valid = false
+        elseif value == 96 || value == 124 || value == 126
+            is_valid = false
+        elseif value == 124
+            is_valid = false
+        else
+            is_valid = true
+        end
```

Comment:
Why not use the character itself. This makes it harder to know what characters you are matching.

![Screen Shot 2020-04-30 at 9 51 50 AM](https://user-images.githubusercontent.com/1813121/80731540-4db4ac80-8ac8-11ea-97ce-4d232b0223ec.png)

Reply:
I considered that.  For the characters that are actually listed, doing that is more readable.  But for the characters not listed it can be harder.  If a reader wants to know what is between `:` and `@` then they have to look at the table.  When looking at the table I find it easier to search for the number.  So, I decided to use numbers everywhere for consistency.

I don't feel strongly about this and will change it if people disagree.  You also seem to be implying that we should check for allowed characters instead of what is not allowed.